CIP-31: updates according to the implemention (#290)
celestiaorg/CIPs: cips/cip-031.md

   Consider an account set up with 1000 tokens scheduled to unlock entirely after 730 days (2 years).
   Assume at day 200, the account holder claims a staking reward of 50 tokens.

   1. The reward is added to the `OriginalVesting` amount tracked by the account: `New Total Original Amount = 1000 + 50 = 1050 tokens`.
   2. Unlike continuous vesting, no portion of this reward unlocks immediately, regardless of the time elapsed.
   3. The account continues to hold 0 unlocked tokens.
   4. When the `EndTime` (day 730) is reached, the *entire* 1050 tokens (original 1000 + 50 reward) unlock at once.

   The claimed rewards simply increase the total balance that will become available at the predefined unlock date.

### Periodic Vesting Accounts

Periodic account creation will be removed, and update schedule will be a noop

### Permanent Vesting Accounts

Permanent account creation will be removed, and update schedule will be a noop

### Module Interaction

The following sequence summarizes the module interactions:

  ```mermaid
  sequenceDiagram
  participant DM as Distribution Module
  participant AK as Account Keeper

>   participant LM as Lockup Module

Reviewer: [nit] it's confusing to call this the lockup module when in code it is the vesting module

Author: i agree, when writing the CIP it was advised to use lockup nomenclature here even though its called vesting. I would prefer to keep it as is. on line 17 I added a note for this confusion.

  participant BM as Bank Module

  DM->>AK: Get Receiving Account
  AK-->>DM: Return Account

  rect rgb(0, 0, 139)
    Note over DM: Account Type Check
    DM->>DM: Is Lockup Account?
  end

  alt Is Lockup Account
    DM->>LM: Get Lockup Type
    DM->>LM: Calculate New Lockup Schedule

>     LM->>LM: Update Lockup Parameters

Reviewer: What does this map to in code?

   - Unlocked & Staked

> 4. **Mathematical Model & Example:**

Reviewer: This list item is no longer needed

```suggestion
```

Author: updated the title but as the previous steps have numbers, breaking it here would cause confusion

Reviewer: If you want a title for the sections then please use a heading level with higher precedence than the H3s that follow. For example

```diff
- 4. **Mathematical Model & Example:**
+ ## Example Implementations

### Continuous Vesting Accounts

### Delayed Vesting Accounts

```

Currently this isn't formatted well. The rendered markdown looks like this

![Screenshot 2025-04-07 at 11 51 02 PM](https://github.com/user-attachments/assets/14196261-707d-4224-8eeb-78e0191b516c)

Author: amended to `#` instead to have a better flow as # and numbers throw things off

   A portion of the reward corresponding to the elapsed vesting duration (30% in this case) unlocks immediately, while the rest (70%) is added to the future vesting schedule.

### Delayed Vesting Accounts

   For a `DelayedVestingAccount`, the entire principal vests at a single point in the future (`EndTime`).
   Consider an account set up with 1000 tokens scheduled to unlock entirely after 730 days (2 years).
   Assume at day 200, the account holder claims a staking reward of 50 tokens.

   1. The reward is added to the `OriginalVesting` amount tracked by the account: `New Total Original Amount = 1000 + 50 = 1050 tokens`.
   2. Unlike continuous vesting, no portion of this reward unlocks immediately, regardless of the time elapsed.
   3. The account continues to hold 0 unlocked tokens.
   4. When the `EndTime` (day 730) is reached, the *entire* 1050 tokens (original 1000 + 50 reward) unlock at once.

   The claimed rewards simply increase the total balance that will become available at the predefined unlock date.

### Periodic Vesting Accounts

Periodic account creation will be removed, and update schedule will be a noop

### Permanent Vesting Accounts

> Periodic account creation will be removed, and update schedule will be a noop

Reviewer: ```suggestion
Permanent account creation will be removed, and update schedule will be a noop
```

>     *NOTE: reward calculation is arbitrary, it is not based on Celestia's inflation.*

Reviewer: this seems out of place
```suggestion
```

> 5. **Error Handling:**
> 
> - Ensure sufficient balance checks before processing rewards.
> - Validate that the lockup period is still active.
> - Separate commission transfers from reward lockup updates.

Reviewer: this isn't relevant because permanent vesting is a no-op. It should be removed from this section. It may be added to another section.

```suggestion
```

# Backwards Compatibility

> Continuous Vesting, Delayed Vesting will implement the `UpdateSchedule` function allowing for the original amount to be added to. Periodic and Permanent will implement the design as a noop, and they will be disallowed from being created. 

Reviewer: ```suggestion
The continuous and delayed vesting account types will be updated to implement CIP-31. Specifically, they will implement the `UpdateSchedule` function so that the original vesting amount can account for staking rewards. The periodic and permanent account types will not support CIP-31. They will be disallowed from being created. 
```

[question] why do we need to disallow creating them? What about periodic and permanent vesting accounts that already exist?

Author: disallowing creation of them in order to simplify the protocol and due to proposing not implementing update schedule. Existing accounts, of which I couldnt find any, would still work, only the creation is disallowed

Reviewer: If existing accounts still work then I don't understand why celestia-app can't support creating new ones. 

Proposal: 
- Implement no-op `UpdateSchedule` for permanent and vesting accounts
- Preserve the ability to create new permanent and vesting accounts
- Clarify CIP-31 only applies to continuous and delayed vesting accounts

Reviewer: Also given the most recent iteration of the implementation takes all staking rewards and updates the original vesting amount, it seems feasible to implement CIP-31 for periodic vesting accounts. AFAIU the complexity in the first implementation occurred when updating the amount to vest for various periods but that logic no longer needs to be added, right?

Author: for periodic this doesnt work as periods are proposed in json on creation. You need to update the periods as the periods contain the amount unlocked.

Reviewer: Got it, thanks! Resolving.

# Test Cases

The implementation should include test cases covering:

- **Basic Reward Claim:**
  - Claim rewards multiple times and verify that the lockup schedule updates correctly.
- **Validator Commission Path:**
  - Confirm that commission rewards are processed normally (non-locked) while staking rewards are added to the lockup schedule.

# Reference Implementation

> The `VestingAccount` interface will be updated to account for an UpdateSchedule method which includes the amount to update the schedule with and teh blocktime at which it occured. We must pass in the blocktime in order to understand if the account is fully vested or not.

Reviewer: ```suggestion
The `VestingAccount` interface will be updated to account for an UpdateSchedule method which includes the amount to update the schedule with and the blocktime at which it occured. We must pass in the blocktime in order to understand if the account is fully vested or not.
```

type VestingAccount interface {
	sdk.AccountI

	// LockedCoins returns the set of coins that are not spendable (i.e. locked),
	// defined as the vesting coins that are not delegated.
	//
	// To get spendable coins of a vesting account, first the total balance must
	// be retrieved and the locked tokens can be subtracted from the total balance.
	// Note, the spendable balance can be negative.
	LockedCoins(blockTime time.Time) sdk.Coins

	// TrackDelegation performs internal vesting accounting necessary when
	// delegating from a vesting account. It accepts the current block time, the
	// delegation amount and balance of all coins whose denomination exists in
	// the account's original vesting balance.
	TrackDelegation(blockTime time.Time, balance, amount sdk.Coins)

	// TrackUndelegation performs internal vesting accounting necessary when a
	// vesting account performs an undelegation.
	TrackUndelegation(amount sdk.Coins)

	GetVestedCoins(blockTime time.Time) sdk.Coins
	GetVestingCoins(blockTime time.Time) sdk.Coins

	GetStartTime() int64
	GetEndTime() int64

	GetOriginalVesting() sdk.Coins
	GetDelegatedFree() sdk.Coins
	GetDelegatedVesting() sdk.Coins

> 	// UpdateSchedule updates the original vesting amount for the account.

Reviewer: ```suggestion
+	// UpdateSchedule updates the original vesting amount for the account.
```
